src/trace_ng/log.zig

```diff
@@ -0,0 +1,273 @@
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        fn withScope(self: Self, comptime new_scope: anytype) StandardLogger(new_scope) {
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.channel.close();
+            if (self.handle) |handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.deinit();
+            self.free_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
+                };
+                defer self.channel.allocator.free(messages);
+
+                for (messages) |message| {
+                    logfmt.writeLog(message) catch @panic("logging failed");
+                }
+            }
+        }
+
+        fn createLogMessage(
+            _: Self,
+            free_fba: *RecycleFBA,
+            total_len: u64,
+            level: Level,
+            maybe_scope: ?[]const u8,
+            maybe_msg: ?[]const u8,
+            maybe_fields: anytype,
+            comptime maybe_fmt: ?[]const u8,
+            args: anytype,
+        ) logfmt.LogMsg {
+            // obtain a memory to write to
+            free_fba.mux.lock();
+            const buf = blk: while (true) {
+                const buf = free_fba.allocator().alloc(u8, total_len) catch {
+                    // no memory available rn - unlock and wait
+                    free_fba.mux.unlock();
+                    std.time.sleep(std.time.ns_per_ms);
+                    free_fba.mux.lock();
+                    continue;
+                };
+                break :blk buf;
+            };
+            free_fba.mux.unlock();
+            errdefer {
+                free_fba.mux.lock();
+                free_fba.allocator().free(buf);
+                free_fba.mux.unlock();
+            }
+            var fmt_message = std.io.fixedBufferStream(buf);
+            const writer = fmt_message.writer();
+
```

Comment:
Doesn't passing `self.fba_bytes` to this function mean that the logger will request an allocation for the entire backing buffer behind the fba? At that point is there any reason to use an allocator instead of passing the entire buffer?

Reply:
For the formatted message, used  `std.fmt.count` to determine how much allocation to ask from the allocator...https://github.com/Syndica/sig/pull/249/files#diff-a5f543f6d9dfe65e2e547cdf7bd2cd49d8e85d4afccc81a3c51568b0208dfeb7R29

For the k/v fields on the other hand, I have not found how to determine the size. Currently hardcoding it to 512 here https://github.com/Syndica/sig/pull/249/files#diff-a5f543f6d9dfe65e2e547cdf7bd2cd49d8e85d4afccc81a3c51568b0208dfeb7R108